File: TypeScript/LeetCode/lc2/solution1.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

 function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let dummyHead = new ListNode(0);

    let carry = 0;
    let node1  = l1;
    let node2 = l2;

    let cur = dummyHead;

    while(node1 != null || node2 != null){
        let v1 = node1 == null ? 0 : node1.val;
        let v2 = node2 == null ? 0 : node2.val;

        cur.next = new ListNode((v1+v2+carry)%10, null);
        cur = cur.next;

        carry = Math.floor((v1+v2+carry)/10);

        if(node1 != null){
            node1 = node1.next;
        }
        if(node2 != null){
            node2 = node2.next;
        }
    }

    if(carry > 0){
        cur.next = new ListNode(carry,null);
    }

    return dummyHead.next;
};

File: Java/LeetCode/from0/lc2/Solution1.java
package lc2;
import java.util.Optional;

import util.ListNode;


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class Solution1 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return carried(l1,l2,0);
    }
    
    public ListNode carried(ListNode l1, ListNode l2, int carry){
        if(l1 == null && l2 == null && carry == 0){
            return null;
        }
        ListNode ret = new ListNode();
        if(l1 != null){
            carry += l1.val;
        }
        if(l2 != null){
            carry += l2.val;
        }
        ret.val = carry % 10;
        ret.next = carried(l1 == null? null : l1.next, l2 == null? null : l2.next, carry /10 );
        return ret;
    } 
}